### PR TITLE
Read design reuse repos and auth token

### PR DESCRIPTION
**Pull request by shrik450**
Relies on the next version of py-allspice being released, so in draft until that goes through.

### CODE REVIEW

**Comment by shrik450** [comment]
(CI expected to fail until the new version of py-allspice is released)

**Review by kdumontnu — COMMENTED** [review]
Minor suggestions. Otherwise, looks great! Pretty straightforward.

File: README.md [hunk 1]
```diff
@@ -203,14 +203,43 @@ columns:
...
-## Group By
+### Design Reuse repositories
+
+For Altium projects, if you are using device sheets that are located in a
+different repository, you can set a list of design reuse repositories. These
+repositories are tried in order to find device sheets that are not in the
+project, i.e. the first repository to contain a file that has the same name as
+a device sheet will be used for that device sheet. If you are using design
+reuse repositories, you MUST set a custom auth token, as the default one may
+not have the permissions to fetch other repositories.
```
[the remark this answers is not in the document]
Comment (shrik450) [reply README.md]: Done - I added a link

**Comment by shrik450** [comment]
Ready to review - the new version is out. I'm also going to set up a couple of test repos that targte this branch to verify that this is working.

**Comment by shrik450** [comment]
Re-requesting review as I had to make a couple of bugfixes. We also now have a test instance here: https://staging.allspice.dev/chaos_engineering/Altium-Hierarchical-Device-Sheet-Repetitions-Demo/actions/runs/5

**Comment by kdumontnu** [comment]
> Re-requesting review as I had to make a couple of bugfixes. We also now have a test instance here: https://staging.allspice.dev/chaos_engineering/Altium-Hierarchical-Device-Sheet-Repetitions-Demo/actions/runs/5

Now that you've put the demo together I would expect the `design_reuse_repos` to be an attribute of the action, not in `columns.yml`. It seems to align with those options a lot more, which are more fundamental, rather than "columns/attributes".

**Comment by shrik450** [comment]
If we put it in the action arguments, we'd run into the same problem that columns.yml was meant to solve: arguments have to be plain strings and cannot be objects or arrays. So it would have to be a newline separated string instead of explicitly an array, and you'd lose the ability to comment them out etc. It's definitely a tradeoff, but I think putting it in the columns.yml file is marginally better (and that is why columns.yml has `columns` as a key instead of it being the whole file!)

**Comment by kdumontnu** [comment]
> If we put it in the action arguments, we'd run into the same problem that columns.yml was meant to solve: arguments have to be plain strings and cannot be objects or arrays. So it would have to be a newline separated string instead of explicitly an array, and you'd lose the ability to comment them out etc. This is more of a problem because the ordering of the repos has meaning. It's definitely a tradeoff, but I think putting it in the columns.yml file is marginally better (and that is why columns.yml has `columns` as a key instead of it being the whole file!)

Augh. It's still not ideal, but I think it's the only path forward right now.

**Comment by shrik450** [comment]
> But either way I guess we would have to build our own parser from strings?

Yes - if we really wanted to keep it in the action arguments, I'd prefer having it as a YAML multiline string, like:

```yml

- uses: ""
  with:
    # ...
    design_reuse_repos: |
        - "owner1/repo1"
        - "owner2/repo2"
```

This way, we wouldn't have to add a parser or write our own, because this action already bundles a YAML parser. For users, it wouldn't be too surprising because they are already writing YAML, and they would still be able to comment/reorder/etc. There are still drawbacks to this: users wouldn't get syntax highlighting and lints if they had something wrong.

**Comment by kdumontnu** [comment]
yeah, I just tested to see what the error would be and it looks like this
image

Not intuitive. I agree, both options have drawbacks. Let's keep it as you have it.

**Review by kdumontnu — APPROVED** [review]
👍

**Pull request merged**